# Exercise 1

In [1]:
import hashlib
import time
import random
import string

In [2]:
# !wget https://raw.githubusercontent.com/danielmiessler/SecLists/master/Passwords/Common-Credentials/10k-most-common.txt

In [3]:
with open('10k-most-common.txt') as f:
    passwords = f.read().split()

print(passwords)

['password', '123456', '12345678', '1234', 'qwerty', '12345', 'dragon', 'pussy', 'baseball', 'football', 'letmein', 'monkey', '696969', 'abc123', 'mustang', 'michael', 'shadow', 'master', 'jennifer', '111111', '2000', 'jordan', 'superman', 'harley', '1234567', 'fuckme', 'hunter', 'fuckyou', 'trustno1', 'ranger', 'buster', 'thomas', 'tigger', 'robert', 'soccer', 'fuck', 'batman', 'test', 'pass', 'killer', 'hockey', 'george', 'charlie', 'andrew', 'michelle', 'love', 'sunshine', 'jessica', 'asshole', '6969', 'pepper', 'daniel', 'access', '123456789', '654321', 'joshua', 'maggie', 'starwars', 'silver', 'william', 'dallas', 'yankees', '123123', 'ashley', '666666', 'hello', 'amanda', 'orange', 'biteme', 'freedom', 'computer', 'sexy', 'thunder', 'nicole', 'ginger', 'heather', 'hammer', 'summer', 'corvette', 'taylor', 'fucker', 'austin', '1111', 'merlin', 'matthew', '121212', 'golfer', 'cheese', 'princess', 'martin', 'chelsea', 'patrick', 'richard', 'diamond', 'yellow', 'bigdog', 'secret', 'as

In [4]:

m=hashlib.md5(b"password").hexdigest()
print(m)
m=hashlib.md5(b"Chulalongkorn").hexdigest()
print(m)
685396
m=hashlib.md5(passwords[0].encode()).hexdigest()
print(m)

finding = "d54cc1fe76f5186380a0939d2fc1723c44e8a5f7"
ch = False

5f4dcc3b5aa765d61d8327deb882cf99
46fa3b56c660faff420190c18c98a56b
5f4dcc3b5aa765d61d8327deb882cf99


In [5]:
def play(idx, st, now):

    if len(passwords[idx]) == st:
        # print(now)
        k = now.encode()
        m = hashlib.md5(k).hexdigest()
        if m == finding:
            print("found :", now)
            return True
        m = hashlib.sha1(k).hexdigest()
        if m == finding:
            print("found :", now)
            return True
        return False
    sub = [['o', '0'], ['l', 'i', '1']]
    key = passwords[idx][st].lower()
    # play(idx, st + 1, now + key)
    chh = True
    ch = False
    for i in sub:
        if key in i:
            chh = False
            for x in i:
                ch = ch | play(idx, st + 1, now + x)
                if x.isalpha():
                    ch = ch | play(idx, st + 1, now + x.upper())
    if chh:
        ch = ch | play(idx, st + 1, now + key)
        if key.isalpha():
            ch = ch | play(idx, st + 1, now + key.upper())
    
    return ch

    

In [6]:
for i in range(len(passwords)):
    if play(i, 0, ""):
        break

found : ThaiLanD


`Answer: ThaiLanD`

# Exercise 2

In [12]:
gen = string.punctuation + string.ascii_letters + string.digits + " "
passwords = []
for i in range(1, 1024+1):
    for k in range(100):
        passwords.append(''.join(random.choices(gen, k=i)))
# print(passwords)
print(len(gen), gen)
# print(len(passwords))

95 !"#$%&'()*+,-./:;<=>?@[\]^_`{|}~abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 


In [8]:
ma = 0
for i in passwords:
    ma = max(ma, len(i))
print(ma)
aa = [0.0 for i in range(1, ma + 2)]
cc = [0 for i in range(1, ma + 2)]
tt = [-1 for i in range(1, ma + 2)]

1024


In [9]:
a = []
t = 0.0
def play2(idx, st, now):
    global t, aa, cc
    if len(passwords[idx]) == st:
        # print(now)
        k = now.encode()
        stt = time.time()
        m = hashlib.sha1(k).hexdigest()
        # n = hashlib.md5(k).hexdigest()
        edt = time.time()
        t += edt - stt
        # print(m, now)
        a.append((m, now))
        aa[st] += edt - stt
        cc[st] += 1
        # a.append((n, now))
        return

    sub = [['o', '0'], ['l', 'i', '1']]
    key = passwords[idx][st].lower()
    chh = True

    for i in sub:
        if key in i:
            chh = False
            for x in i:
                play2(idx, st + 1, now + x)
                if x.isalpha():
                    play2(idx, st + 1, now + x.upper())
    if chh:
        play2(idx, st + 1, now + key)
        if key.isalpha():
            play2(idx, st + 1, now + key.upper())

    # play2(idx, st + 1, now + key)

    return

st = time.time()
for i in range(len(passwords)):
    play2(i, 0, "")
ed = time.time()

In [56]:
print("total :", len(a), "items")
print("total create time :", ed - st, "sec")

total : 13685396 items
total create time : 9.084963321685791 sec


`time for create is 9.08 sec and table size is 13685396 items`

# Exercise 3

In [85]:
print("total hash time :", t, "sec")
print("Avg time per hash :", t*1e9/len(a), "ns")

total hash time : 4.120308876037598 sec
Avg time per hash : 301.0734125660374 ns


In [11]:
for i in range(len(aa)):
    if cc[i] != 0:
        tt[i] = aa[i] * 1e9 / cc[i]
    print(i, tt[i])
# print(tt)

0 -1
1 360.0120544433594
2 278.9497375488281
3 300.40740966796875
4 314.7125244140625
5 302.7915954589844
6 314.7125244140625
7 295.6390380859375
8 329.01763916015625
9 309.94415283203125
10 317.0967102050781
11 293.2548522949219
12 331.4018249511719
13 302.7915954589844
14 333.7860107421875
15 317.0967102050781
16 329.01763916015625
17 352.8594970703125
18 455.3794860839844
19 360.0120544433594
20 357.62786865234375
21 331.4018249511719
22 362.396240234375
23 338.55438232421875
24 331.4018249511719
25 321.8650817871094
26 336.1701965332031
27 431.5376281738281
28 376.70135498046875
29 343.32275390625
30 333.7860107421875
31 329.01763916015625
32 352.8594970703125
33 393.3906555175781
34 326.6334533691406
35 391.0064697265625
36 443.45855712890625
37 374.3171691894531
38 371.9329833984375
39 350.4753112792969
40 355.2436828613281
41 374.3171691894531
42 367.16461181640625
43 452.99530029296875
44 343.32275390625
45 452.99530029296875
46 352.8594970703125
47 393.3906555175781
48 360.012

`For not above 32 bytes hashing, it's take nearly time for sha1 hashing from the experiment. For 13685396 items take 4.12 secs for hashing which is around 301 ns per hash. (I use 4.12 which is only hashing time because python has too many overheads and let's assume C++ has nearly 0 overheads)`

# Exercise 4

\begin{align*}
\text{Time For Crack} = K^N \cdot T \\
&\text{Where,} \\
&\text{K = Number of possible characters} \\
&\text{N = Length of password} \\
&\text{T = Time for each hashing}
\end{align*}

# Exercise 5

`Let's`
- `K = 95 (26: a-z), (26: A-Z), (10: 0-9), (32: special characters), (1: space)`
- `T = 301 ns`

For `4 characters password` take 95^4 * 301 ns = `around 24.5 seconds`

For `5 characters password` take 95^5 * 301 ns = `around 39 minutes`

For `6 characters password` take 95^6 * 301 ns = `around 2 and a half days`

For `7 characters password` take 95^7 * 301 ns = `around 8 months`

For `8 characters password` take 95^8 * 301 ns = `around 63 years`

For `9 characters password` take 95^9 * 301 ns = `around 6011 years`

Let's say I might not change the password to the end of my life, so I can use 8 characters password or 9 characters is plus for more security.

# Exercise 6

For each password will have corresponding salt and use it along with the password to hash. So, the hash will be different for the same password.  
ex. `password = "1234"` and `salt = "@#_@!"` will hash to `hash("@#_@!1234")`

1. Password that pass to hash is now longer which harder to crack.

2. Even if the password is as same as another user, the salt will be different make rainbow table useless due to different hash for the same password.